server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/Requeue.java

```diff
@@ -0,0 +1,118 @@
+/**
+ * <p><b>Requeue</b> puts back the email in a queue.
+ * It can be used for throttling when combined with rate limiting mailet. {@link org.apache.james.transport.mailets.PerSenderRateLimit}</p>
+ *
+ * <ul>Here are supported configuration parameters:
+ *      <li><b>queue</b>: a Mail Queue name (optional, default to spool).</li>
+ *      <li><b>processor</b>: a target processor (optional, defaults to root).</li>
+ *      <li><b>delay</b>: a delay when en-queueing mail (optional, defaults to none). Supported units include: s (second), m (minute), h (hour), d (day).</li>
+ *      <li><b>consume</b>: a processor to which current email should be redirected to (optional, defaults to ghost).</li>
+ *  </ul>
+ *
+ *  <p>For instance, to apply all the examples given above:</p>
+ *
+ *  <pre><code>
+ *  &lt;mailet matcher=&quot;All&quot; class=&quot;Requeue&quot;&gt;
+ *      &lt;queue&gt;spool&lt;/queue&gt;
+ *      &lt;processor&gt;root&lt;/processor&gt;
+ *      &lt;delay&gt;2h&lt;/delay&gt;
+ *      &lt;consume&gt;ghost&lt;/consume&gt;
+ *  &lt;/mailet&gt;
+ *   </code></pre>
+ *
+ * <p>Note that this is a naive approach: if you have a large number of emails (say N) submitted at once,
+ * you would need O(N) re-queues to eventually send all your emails,
+ * where the constant is the number of mail allowed over the time window.
+ * So this gives an overall complexity of O(N2), while careful planning could achieve this in O(N). </p>
+ */
+public class Requeue extends GenericMailet {
+    private final MailQueueFactory<?> mailQueueFactory;
+
+    private MailQueue mailQueue;
+    private Optional<Duration> delayDuration;
+    private String processor;
+    private String consumeProcessor;
+
+    /**
+     *
+     * @param mailQueueFactory Allows to get mail queue by name.
+     */
+    @Inject
+    public Requeue(MailQueueFactory<?> mailQueueFactory) {
+        this.mailQueueFactory = mailQueueFactory;
+    }
+
+    @Override
+    public void init() throws MessagingException {
+        MailQueueName mailQueueName = Optional.ofNullable(getInitParameter("queue"))
+            .map(MailQueueName::of).orElse(MailQueueFactory.SPOOL);
+
+        mailQueue = mailQueueFactory.getQueue(mailQueueName)
+            .orElseThrow(() -> new IllegalArgumentException("Can not find queue " + mailQueueName.asString()));
+        delayDuration = Optional.ofNullable(getInitParameter("delay"))
+            .map(delayValue -> DurationParser.parse(delayValue, ChronoUnit.SECONDS));
+        processor = Optional.ofNullable(getInitParameter("processor"))
+            .orElse(Mail.DEFAULT);
+        consumeProcessor = Optional.ofNullable(getInitParameter("consume"))
+            .orElse(Mail.GHOST);
+
+        Preconditions.checkArgument(delayDuration.isEmpty() || !delayDuration.get().isNegative(),
+            "Duration should be non-negative");
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        mail.setState(consumeProcessor);
```

Comment:
```suggestion
if (consume) {
   enqueue(mail);
   mail.setState(GHOST);
} else {
        Mail newMail = null;
        try {
            newMail = mail.duplicate();
            enqueue(newMail)
        } finally {
            LifecycleUtil.dispose(newMail);
        }
} 
```

With

```
private void enqueue(Mail mail) {
            mail.setState(processor);
            duration.ifPresentOrElse(
                delay -> mailQueue.enQueue(mail, delay),
                () -> mailQueue.enQueue(mail));
}
```

 - 1. Don't duplicate the mail when consume is true: that's expensive
 - 2. If spotted on an Optional which is an anti-pattern
 - 3. Correct the handling of consume...

Reply:
if consume = false, mail will be sent duplicate. 
Does it make sense? While The purpose of this mailet is "throttling"?

Reply:
> Does it make sense?

Yes, if you want for some reason to duplicate the mail.

>  While The purpose of this mailet is "throttling"?

The purpose of this mailet is "Requeue". Period. It can be used for throttling but it is just one of the many possible usages.